pyrefly/lib/alt/class/class_field.rs

```diff
@@ -1782,4 +1782,117 @@ impl<'a, Ans: LookupAnswer> AnswersSolver<'a, Ans> {
                 .and_then(|ty| make_bound_method(Type::type_form(cls.clone().to_type()), ty).ok())
         }
     }
+
+    /// Check if a class inherits from abc.ABC
+    ///
+    /// This recursively traverses the class hierarchy to determine if the class
+    /// inherits from Python's abstract base class (abc.ABC). Only classes that
+    /// inherit from abc.ABC can be considered abstract in Python.
+    fn inherits_from_abc(&self, cls: &ClassType) -> bool {
+        let class_obj = cls.class_object();
+
+        // Check if this class is abc.ABC itself
+        if class_obj.module_name().as_str() == "abc" && class_obj.name().as_str() == "ABC" {
+            return true;
+        }
+
+        // Recursively check base classes
+        let metadata = self.get_metadata_for_class(class_obj);
+        for base in metadata.bases_with_metadata() {
+            if self.inherits_from_abc(&base.0) {
+                return true;
+            }
+        }
+
+        false
+    }
+
+    /// Check if a class has any unimplemented abstract methods
+    ///
+    /// Returns true if the class:
+    /// 1. Inherits from abc.ABC (or is abc.ABC itself)
+    /// 2. Has abstract methods in its hierarchy that are not implemented
+    ///
+    /// This method is used to prevent instantiation of abstract classes.
+    pub fn has_abstract_methods(&self, cls: &ClassType) -> bool {
+        // Only classes that inherit from abc.ABC can be abstract
+        if !self.inherits_from_abc(cls) {
+            return false;
+        }
+
+        // Check for unimplemented abstract methods with early exit
+        self.has_unimplemented_abstract_methods(cls)
+    }
+
+    /// Check if a class has unimplemented abstract methods
+    fn has_unimplemented_abstract_methods(&self, cls: &ClassType) -> bool {
+        let mut abstract_methods = SmallSet::new();
+
+        // Collect all abstract methods from the hierarchy
+        self.collect_abstract_methods_from_class(cls, &mut abstract_methods);
+
+        // Check each abstract method for concrete implementation
+        for method_name in &abstract_methods {
+            if !self.class_provides_concrete_implementation(cls, method_name) {
+                return true;
+            }
+        }
+
+        false
+    }
+
+    /// Recursively collect abstract methods from a class and its bases
+    fn collect_abstract_methods_from_class(
```

Comment:
I'm not sure we need to traverse all the base classes like this. If we look up each field the class has like so https://github.com/facebook/pyrefly/blob/b35b7ab29e154af6cb7acaad495afd855e1d33ef/pyrefly/lib/alt/class/class_field.rs#L1589 it would handle the lookup, and we can check if the field it returns is abstract or not.

Of course, this relies on the MRO being correct and able to handle cases when one base class declares something as abstract and another base class provides the impl. @stroxler would know more about this